File: 0020_Valid_Parentheses/Solution.java
public class Solution {
    public boolean isValid(String s) {
        char [] stack = new char[s.length()];
        int head = 0;
        for (int i = 0; i < s.length(); ++i) {
        	char c = s.charAt(i);
        	switch(c) {
        		case '(':
        		case '{':
        		case '[':
        		    stack[head++] = c;
        		    break;
        		case ')':
        		    if (head == 0 || stack[--head] != '(') return false;
        		    break;
        		case '}':
        		    if (head == 0 || stack[--head] != '{') return false;
        		    break;
        		case ']':
        		    if (head == 0 || stack[--head] != '[') return false;
        		    break;
        	}
        }
        return head == 0;
    }
}

File: 0020_Valid_Parentheses/solution.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []
        matches = {'(': ')', '[': ']', '{': '}'}
        for c in s:
        	if c in matches:
        		stack.append(c)
        	elif len(stack) == 0 or matches[stack.pop()] != c:
        		return False
        return len(stack) == 0
